lib/block_view_message.go

```diff
@@ -79,86 +178,172 @@ func (bav *UtxoView) deleteMessageMappings(message *PGMessage) {
 
-	// Start by fetching all the messages we have in the db.
-	dbMessageEntries, err := DbGetLimitedMessageEntriesForPublicKey(bav.Handle, publicKey)
+	// First get all messaging keys for a user.
+	var err error
+	_messagingKeyEntries, err = bav.GetUserMessagingKeys(publicKey)
 	if err != nil {
-		return nil, errors.Wrapf(err, "GetMessagesForUser: Problem fetching MessageEntrys from db: ")
+		return nil, nil, errors.Wrapf(err, "GetLimitedMessagesForUser: " +
+			"problem getting user messaging keys")
 	}
 
-	// Iterate through the entries found in the db and force the view to load them.
-	// This fills in any gaps in the view so that, after this, the view should contain
-	// the union of what it had before plus what was in the db.
-	for _, dbMessageEntry := range dbMessageEntries {
-		messageKey := MakeMessageKey(publicKey, dbMessageEntry.TstampNanos)
-		bav._getMessageEntryForMessageKey(&messageKey)
+	// We define an auxiliary map to keep track of messages in UtxoView and DB.
+	messagesMap := make(map[MessageKey]*MessageEntry)
+
+	// First look for messages in the UtxoView. We don't skip deleted entries for now as we will do it later.
+	for messageKey, messageEntry := range bav.MessageKeyToMessageEntry {
+		for _, messagingKeyEntry := range _messagingKeyEntries {
+			if reflect.DeepEqual(messageKey.PublicKey[:], messagingKeyEntry.MessagingPublicKey[:]) {
+				// We will add the messages by sender so that we have no overlaps in the DB in some weird edge cases.
+				mapKey := MakeMessageKey(messageEntry.SenderMessagingPublicKey[:], messageEntry.TstampNanos)
+				messagesMap[mapKey] = messageEntry
+				break
+			}
+		}
 	}
 
-	// Now that the view mappings are a complete picture, iterate through them
-	// and set them on the map we're returning. Skip entries that don't match
-	// our public key or that are deleted. Note that only considering mappings
-	// where our public key is part of the key should ensure there are no
-	// duplicates in the resulting list.
-	messageEntriesToReturn := []*MessageEntry{}
-	for viewMessageKey, viewMessageEntry := range bav.MessageKeyToMessageEntry {
-		if viewMessageEntry.isDeleted {
-			continue
-		}
-		messageKey := MakeMessageKey(publicKey, viewMessageEntry.TstampNanos)
-		if viewMessageKey != messageKey {
-			continue
+	// Now look for messages in the DB.
+	dbMessageEntries, err := DbGetLimitedMessageForMessagingKeys(bav.Handle, _messagingKeyEntries, limit)
+	if err != nil {
+		return nil, nil, errors.Wrapf(err, "GetMessagesForUser: Problem fetching MessageEntries from db: ")
+	}
+	for _, messageEntry := range dbMessageEntries {
+		mapKey := MakeMessageKey(messageEntry.SenderMessagingPublicKey[:], messageEntry.TstampNanos)
+		if _, exists := messagesMap[mapKey]; !exists {
+			messagesMap[mapKey] = messageEntry
+		} else {
+			// If the message was deleted in the UtxoView, then we remove it from our map struct.
+			if messagesMap[mapKey].isDeleted {
+				delete(messagesMap, mapKey)
+			}
```

Comment:
Once again, this is a little concerning. We typically don't delete entries that are marked with `isDeleted=true` when we're reading from the view

Reply:
Should be fineee, per above comment.